src/interpreter/ByteCodeGenerator.cpp

```diff
@@ -88,14 +88,34 @@ void ByteCodeGenerateContext::morphJumpPositionIntoComplexCase(ByteCodeBlock* cb
 }
 
 #ifdef ESCARGOT_DEBUGGER
-void ByteCodeGenerateContext::insertBreakpoint(size_t line, Node* node)
+void ByteCodeGenerateContext::insertBreakpoint(size_t index, Node* node)
 {
     ASSERT(m_breakpointContext != nullptr);
+    ASSERT(index != SIZE_MAX);
+
+    StringView src = m_codeBlock->asInterpretedCodeBlock()->src();
+    ExtendedNodeLOC sourceElementStart = m_codeBlock->asInterpretedCodeBlock()->functionStart();
+    size_t lastLineOffset = m_breakpointContext->m_lastBreakpointLineOffset;
+    index -= sourceElementStart.index;
+
+    if (index > m_breakpointContext->m_lastBreakpointIndexOffset) {
+        for (size_t i = m_breakpointContext->m_lastBreakpointIndexOffset; i < index; i++) {
+            char16_t c = src.charAt(i);
+            if (EscargotLexer::isLineTerminator(c)) {
+                // skip \r\n
+                if (UNLIKELY(c == 13 && (i + 1 < index) && src.charAt(i + 1) == 10)) {
+                    i++;
+                }
+                m_breakpointContext->m_lastBreakpointLineOffset++;
+            }
+        }
+
+        m_breakpointContext->m_lastBreakpointIndexOffset = index;
+    }
 
-    if (line != 0 && line != m_breakpointContext->m_lastBreakpointLine) {
-        m_breakpointContext->m_breakpointLocations.push_back(Debugger::BreakpointLocation(line, (uint32_t)m_byteCodeBlock->currentCodeSize()));
+    if (lastLineOffset == 0 || lastLineOffset != m_breakpointContext->m_lastBreakpointLineOffset) {
```

Comment:
How about simplifying this check code like `lastLine > m_breakpointContext->m_lastBreakpointLine`?
For the first breakpoint insertion, we can also handle this corner case by setting the initial value.
For example,
```suggestion
    size_t lastLine = m_breakpointContext->m_lastBreakpointLine == 0 ? sourceElementStart.line : m_breakpointContext->m_lastBreakpointLine;
    ...
    if (lastLine > m_breakpointContext->m_lastBreakpointLine) {
```